packages/astro/src/core/app/index.ts

```diff
@@ -250,12 +253,53 @@ export class App {
 		return pathname;
 	}
 
+	#redirectTrailingSlash(pathname: string): string | undefined {
+		const { trailingSlash } = this.#manifest;
+
+		// Ignore root and internal paths
+		if (pathname === '/' || pathname.startsWith('/_')) {
+			return;
+		}
+
+		// Redirect multiple trailing slashes to collapsed path
+		if (MANY_SLASHES.test(pathname)) {
+			const path = pathname.replace(MANY_SLASHES, trailingSlash === 'never' ? '' : '/');
+			return path ? path : '/';
+		}
+
+		if (trailingSlash === 'ignore') {
+			return;
+		}
+
+		// Add a trailing slash if it's not a file
+		const hasTrailingSlash = pathname.endsWith('/');
+		if (trailingSlash === 'always' && !hasTrailingSlash && !hasFileExtension(pathname)) {
+			return `${pathname}/`;
+		}
+		if (trailingSlash === 'never' && hasTrailingSlash) {
+			return pathname.slice(0, -1);
+		}
```

Comment:
We already have a bunch of methods to handle trailing slashes, let's use them instead of re-inventing the wheel: 
- `prependForwardSlash`
- `removeTrailingForwardSlash`